File: examples/python-interop-v0.1/verify_receipt.py
#!/usr/bin/env python3
"""Minimal BoundaryAttest Interop Profile v0.1 receipt verifier."""

import argparse
import base64
import binascii
import hashlib
import json
from pathlib import Path

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

from sign_receipt import REQUIRED_CLAIM_FIELDS, canonical_claim_bytes


TOP_LEVEL_FIELDS = ("claim", "signature", "public_key_id")


def fail(reason: str) -> dict[str, object]:
    return {"ok": False, "reason": reason}


def verify_receipt(receipt_text: str, public_key_pem: bytes) -> dict[str, object]:
    try:
        receipt = json.loads(receipt_text)
    except (json.JSONDecodeError, ValueError):
        return fail("invalid_json")
    if not isinstance(receipt, dict):
        return fail("invalid_receipt")
    for field in TOP_LEVEL_FIELDS:
        if field not in receipt:
            return fail(f"missing_top_level_field:{field}")
    for field in receipt:
        if field not in TOP_LEVEL_FIELDS:
            return fail(f"unexpected_top_level_field:{field}")
    claim = receipt["claim"]
    if not isinstance(claim, dict):
        return fail("claim_not_object")
    if not isinstance(receipt["signature"], str) or not isinstance(
        receipt["public_key_id"], str
    ):
        return fail("invalid_receipt")
    for field in REQUIRED_CLAIM_FIELDS:
        if field not in claim:
            return fail(f"missing_claim_field:{field}")
    if claim["receipt_version"] != "0.1":
        return fail("unsupported_version")
    if claim["receipt_role"] not in ("client_observed", "server_attested"):
        return fail("unsupported_receipt_role")

    try:
        public_key = serialization.load_pem_public_key(public_key_pem)
        if not isinstance(public_key, Ed25519PublicKey):
            return fail("public_key_id_mismatch")
        public_der = public_key.public_bytes(
            serialization.Encoding.DER, serialization.PublicFormat.SubjectPublicKeyInfo
        )
    except (TypeError, ValueError):
        return fail("public_key_id_mismatch")
    expected_key_id = "sha256:" + hashlib.sha256(public_der).hexdigest()
    if receipt["public_key_id"] != expected_key_id:
        return fail("public_key_id_mismatch")

    try:
        signature = base64.b64decode(receipt["signature"], validate=True)
        public_key.verify(signature, canonical_claim_bytes(claim))
    except (InvalidSignature, ValueError, TypeError, binascii.Error):
        return fail("invalid_signature")
    return {"ok": True}


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("receipt", type=Path)
    parser.add_argument("public_key", type=Path, help="expected Ed25519 SPKI PEM public key")
    args = parser.parse_args()
    result = verify_receipt(
        args.receipt.read_text(encoding="utf-8"), args.public_key.read_bytes()
    )
    print(json.dumps(result, separators=(",", ":")))
    raise SystemExit(0 if result["ok"] else 1)


if __name__ == "__main__":
    main()

File: examples/interop-v0.2/verify-receipt.ts
import { createHash, createPublicKey, sign, verify } from "node:crypto";
import { jcsCanonicalBytes } from "../../src/jcs.js";

const TOP_LEVEL_FIELDS = ["claim", "signature", "public_key_id"] as const;
export const REQUIRED_CLAIM_FIELDS = ["receipt_version", "receipt_role", "event_id", "timestamp", "action_type", "status"] as const;
export type FailureReason = "invalid_json" | "invalid_receipt" | `missing_top_level_field:${string}` | `unexpected_top_level_field:${string}` | "claim_not_object" | `missing_claim_field:${string}` | "unsupported_version" | "unsupported_receipt_role" | "public_key_id_mismatch" | "invalid_signature";
export type VerificationResult = { ok: true } | { ok: false; reason: FailureReason };

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}
const hasOwn = (value: object, field: string): boolean => Object.prototype.hasOwnProperty.call(value, field);

export function interopV02PublicKeyId(publicKeyPem: string): string {
  const der = createPublicKey(publicKeyPem).export({ type: "spki", format: "der" });
  return `sha256:${createHash("sha256").update(der).digest("hex")}`;
}

export function signInteropV02Claim(claim: Record<string, unknown>, privateKeyPem: string): string {
  return sign(null, jcsCanonicalBytes(claim), privateKeyPem).toString("base64");
}

export function verifyInteropV02Receipt(receiptText: string, publicKeyPem: string): VerificationResult {
  let receipt: unknown;
  try { receipt = JSON.parse(receiptText); } catch { return { ok: false, reason: "invalid_json" }; }
  if (!isRecord(receipt)) return { ok: false, reason: "invalid_receipt" };
  for (const field of TOP_LEVEL_FIELDS) if (!hasOwn(receipt, field)) return { ok: false, reason: `missing_top_level_field:${field}` };
  for (const field of Object.keys(receipt)) if (!(TOP_LEVEL_FIELDS as readonly string[]).includes(field)) return { ok: false, reason: `unexpected_top_level_field:${field}` };
  if (!isRecord(receipt.claim)) return { ok: false, reason: "claim_not_object" };
  if (typeof receipt.signature !== "string" || typeof receipt.public_key_id !== "string") return { ok: false, reason: "invalid_receipt" };
  for (const field of REQUIRED_CLAIM_FIELDS) if (!hasOwn(receipt.claim, field)) return { ok: false, reason: `missing_claim_field:${field}` };
  if (receipt.claim.receipt_version !== "0.2") return { ok: false, reason: "unsupported_version" };
  if (receipt.claim.receipt_role !== "client_observed" && receipt.claim.receipt_role !== "server_attested") return { ok: false, reason: "unsupported_receipt_role" };
  let keyId: string;
  try { keyId = interopV02PublicKeyId(publicKeyPem); } catch { return { ok: false, reason: "public_key_id_mismatch" }; }
  if (receipt.public_key_id !== keyId) return { ok: false, reason: "public_key_id_mismatch" };
  try {
    return verify(null, jcsCanonicalBytes(receipt.claim), publicKeyPem, Buffer.from(receipt.signature, "base64"))
      ? { ok: true }
      : { ok: false, reason: "invalid_signature" };
  } catch { return { ok: false, reason: "invalid_signature" }; }
}
